src/plans/vectorial_plan.jl

```diff
@@ -0,0 +1,520 @@
+
+# Fields
+
+* `value!!`:          the cost function ``f``, which can take different formats
+* `cost_type`:     indicating / string data for the type of `f`
+* `jacobian!!:     the jacobian of ``f``
+* `jacobian_type`: indicating / storing data for the type of ``J_f``
+* `parameters`:    the nunmber `n` from above, that is the size of the vector ``f`` returns.
+
+# Constructor
+
+VectorGradientFunction(f, Jf, range_dimension;
+    evaluation::AbstractEvaluationType=AllocatingEvaluation(),
+    function_type::AbstractVectorialType=FunctionVectorialType(),
+    jacobian_type::AbstractVectorialType=FunctionVectorialType(),
+)
+"""
+struct VectorGradientFunction{
+    E<:AbstractEvaluationType,
+    FT<:AbstractVectorialType,
+    JT<:AbstractVectorialType,
+    F,
+    J,
+    I<:Integer,
+} <: Function
+    value!!::F
+    cost_type::FT
+    jacobian!!::J
+    jacobian_type::JT
+    range_dimension::I
+end
+
+function VectorGradientFunction(
+    f::F,
+    Jf::J,
+    range_dimension::I;
+    evaluation::E=AllocatingEvaluation(),
+    function_type::FT=FunctionVectorialType(),
+    jacobian_type::JT=FunctionVectorialType(),
+) where {
+    I,F,J,E<:AbstractEvaluationType,FT<:AbstractVectorialType,JT<:AbstractVectorialType
+}
+    return VectorGradientFunction{E,FT,JT,F,J,I}(
+        f, function_type, Jf, jacobian_type, range_dimension
+    )
+end
+
+@doc raw"""
+    get_value(M::AbstractManifold, vgf::VectorGradientFunction, p, i, range=nothing)
+
+Evaluate the vector function [`VectorGradientFunction`](@ref) `vgf` at at `p`.
+The `range` can be used to speficy a potential range, but is currently only present for consistency.
+
+The `i` can be a linear index, you can provide
+
+* a single integer
+* a `UnitRange` to specify a range to be returned like `1:3`
+* a `BitVector` specifying a selection
+* a `AbstractVector{<:Integer}` to specify indices
+* `:` to return the vector of all gradients
+
+"""
+get_value(M::AbstractManifold, vgf::VectorGradientFunction, p, i; range=nothing)
+function get_value(
+    M::AbstractManifold,
+    vgf::VectorGradientFunction{E,<:FunctionVectorialType},
+    p,
+    i,
+    range=nothing,
+) where {E}
+    c = vgf.value!!(M, p)
+    if isa(c, Number)
+        return c
+    else
+        return c[i]
+    end
+end
+function get_value(
+    M::AbstractManifold,
+    vgf::VectorGradientFunction{E,<:ComponentVectorialType},
+    p,
+    i::Integer,
+    range=nothing,
+) where {E}
+    return vgf.value!![i](M, p)
+end
+function get_value(
+    M::AbstractManifold,
+    vgf::VectorGradientFunction{E,<:ComponentVectorialType},
+    p,
+    i,
+    range=nothing,
+) where {E}
+    return [f(M, p) for f in vgf.value!![i]]
+end
+
+@doc raw"""
+    get_value_function(vgf::VectorGradientFunction, recursive=false)
+
+return the internally stored function computing [`get_value`](@ref).
+"""
+function get_value_function(vgf::VectorGradientFunction, recursive=false)
```

Comment:
What is `recursive` for?

Reply:
If you have multiple wrappers, whether you just want to remove one or continue. All function access functions have this.

Reply:
I see, OK.

Reply:
Hm, probably too much work to write a `CacheFunction` thingy just for the vgf, then I will keep the double-code cache stuff for eq and inequality constraints for now and rework that to properly work with more than integer `j`.